### ISSUE TITLE
"Flow Cancelled" Exception When Activity is Destroyed

### ISSUE THREAD

**Issue by noelstieglitz** [okta/okta-mobile-kotlin#248]
### Describe the bug?

When the activity is destroyed due to configuration change, such as device rotation or entering multi-window mode, etc. and the user attempts to login or logout using the chrome custom tab, the user will not be able to complete the operation and the SDK will provide the following exception:

```
Flow cancelled.
                                                                                                    com.okta.webauthenticationui.WebAuthenticationClient$FlowCancelledException: Flow cancelled.
                                                                                                    	at com.okta.webauthenticationui.DefaultRedirectCoordinator.emit(RedirectCoordinator.kt:123)
                                                                                                    	at com.okta.webauthenticationui.SingletonRedirectCoordinator.emit(Unknown Source:2)
                                                                                                    	at com.okta.webauthenticationui.ForegroundViewModel.flowCancelled(ForegroundViewModel.kt:78)
                                                                                                    	at com.okta.webauthenticationui.ForegroundActivity.onPause(ForegroundActivity.kt:67)
                                                                                                    	at android.app.Activity.performPause(Activity.java:8481)
                                                                                                    	at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1618)
                                                                                                    	at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:5054)
                                                                                                    	at android.app.ActivityThread.performPauseActivity(ActivityThread.java:5015)
                                                                                                    	at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4967)
                                                                                                    	at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:48)
                                                                                                    	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7872)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

### What is expected to happen?

The user is able to login and logout.

### What is the actual behavior?

The SDK throws a `com.okta.webauthenticationui.WebAuthenticationClient$FlowCancelledException`.

### Reproduction Steps?

1. In the Android developer tools, toggle "Don't keep activities" to on. 
2. Attempt to login using the web authentication client

**Actual Results**
The SDK throws a `com.okta.webauthenticationui.WebAuthenticationClient$FlowCancelledException`.

**Expected Results**
The SDK retains state so the user is able to login (perhaps using `onSaveInstanceState`).

### Additional Information?

I suspect this is an issue for multi window devices as well, based on production reports.  Unfortunately I cannot attempt to reproduce.

### SDK Version and Artifact(s) used.

1.12

**Comment by rajdeepnanua-okta** [comment]
Hi @noelstieglitz,

Thanks for reporting this bug. I will look into this.

**Comment by rajdeepnanua-okta** [comment]
Creating a fix for this issue is non-trivial. Ideally, receiving browser callback shouldn't depend on the activity lifecycle at all. There are 2 issues that connect WebAuthenticationClient.login's functionality with ForegroundActivity's lifecycle:
1. We cancel the flow when ForegroundActivity is finishing. This one is the easier problem to fix. But, after fixing this, there's another harder problem,
2. When ForegroundActivity is destroyed and we receive the browser callback, ForegroundActivity is created again along with ForegroundViewModel. This causes our SDK to not return any result to the login function.

I need to re-evaluate how the redirect result from ForegroundActivity can be passed back more reliably to login function call without depending on ForegroundActivity's lifecycle. Internal issue reference: OKTA-604954

**Comment by noelstieglitz** [comment]
@rajdeepnanua-okta  thanks for the detailed update. I’m guessing that, since the fix will be non-trivial it may take a while to get a fix merges. I suspect this issue is affecting a decent number of users based on negative app reviews as well as logs showing many flow cancellation exceptions. With that,

 * Do you have any suggestions on a temporary workaround?
 * It could be that we are seeing the cancelled flow exceptions for a variety of reasons.  I don’t see anyway to different between the myriad of reasons a user may see that exception. Is there anyway more granular details could be exposed via the error response to indicate why the exception occurred?  This is a lower priority and really a separate issue for me. Perhaps I should open a different GitHub issue with more details?

**Comment by rajdeepnanua-okta** [comment]
Hi @noelstieglitz, I have fixed the issue with login flow being incorrectly cancelled on activity onDestroy. With the fix, there are a few caveats that need to be addressed in the caller of webAuthenticationClient.login(). ViewModel might get destroyed right after Activity.onDestroy, so that can stop the coroutines running in viewModelScope and also make the state holder in that viewmodel inaccessible. So, you might need to use coroutine scope and data holders which survive viewmodel.onClear.

Also, I discovered another caveat related to "Don't keep activities" debug option. On some devices, turning that option on causes previously destroyed activity to show up when the browser redirects to the SDK with login credentials, and then another activity is created as well. It's not possible for the SDK to tell if the user cancelled the login flow by pressing the close button, or if we run into this scenario. Pixel 2 API 33 emulator seems to not have this issue.

I looked into improving logging as well, but it's not easy to tell from our end whether the user quit the login flow by pressing the close button, or if we throw the error without user interaction. The browser window unfortunately doesn't give us this info, so we assume that if our activity is resumed and we didn't get new login data, then the user cancelled the flow.

I will create a new release with this fix today.
